### PR TITLE
feat: add more rocksdb  perf-counter support

### PR DESCRIPTION
**Pull request by Shuo-Jia**
### What problem does this PR solve? 
some rocksdb has influence with read/write latency, we need add perf-counter support

##### Tests 

- Manual test 

##### Side effects
read amp need external memory to obtain, detail see: https://github.com/XiaoMi/pegasus-rocksdb/blob/v6.6.4-compatible/include/rocksdb/table.h#L247, **so we disable it by default**

### Config update
```diff
+ read_amp_bytes_per_bit = 0 # 0 means disble read amp counter
- update_rdb_stat_interval = 600
+ update_rdb_stat_interval = 60
```
### Counter update
```diff
+ replica*app.pegasus*rdb.read_amplification@gpid
+ replica*app.pegasus*rdb.write_amplification@gpid
+ replica*app.pegasus.rdb.read_memtable_total_count@gpid
+ replica*app.pegasus.rdb.read_memtable_hit_count@gpid
+ replica*app.pegasus*rdb.read_l0_hit_count@gpid
+ replica*app.pegasus*rdb.read_l1_hit_count@gpid
+ replica*app.pegasus*rdb.read_l2andup_hit_count@gpid
```

### CODE REVIEW

File: src/server/pegasus_server_impl.cpp [hunk 1]
```diff
@@ -1532,13 +1535,20 @@ ::dsn::error_code pegasus_server_impl::start(int argc, char **argv)
                                       [this]() { this->update_replica_rocksdb_statistics(); },
                                       _update_rdb_stat_interval);
 
+    dinfo_replica("start the update rocksdb amp statistics timer task");
+    _update_replica_amp_stat =
+        ::dsn::tasking::enqueue_timer(LPC_REPLICATION_LONG_COMMON,
+                                      nullptr,
```
Comment (Smityz) [inline src/server/pegasus_server_impl.cpp:1541]: why not use tracker here

File: src/server/pegasus_server_impl.h [hunk 2]
```diff
@@ -39,7 +40,7 @@
 
 namespace pegasus {
 namespace server {
-
+DSN_DECLARE_int32(read_amp_bytes_per_bit);
```
Comment (levy5307) [inline src/server/pegasus_server_impl.h:43]: move this to the corresponding cpp file

File: src/server/pegasus_server_impl_init.cpp [hunk 3]
```diff
@@ -44,6 +44,20 @@ DSN_DEFINE_bool("pegasus.server",
...
+// https://github.com/XiaoMi/pegasus-rocksdb/blob/v6.6.4-compatible/include/rocksdb/table.h#L247
+DSN_DEFINE_int32("pegasus.server",
+                 read_amp_bytes_per_bit,
+                 0,
+                 "config for using to calculate the "
+                 "read amplification, must be a power "
+                 "of 2, zero means disable count read "
+                 "amplification");
+
+DSN_DEFINE_validator(read_amp_bytes_per_bit, [](const int64_t read_amp_bytes_per_bit) -> bool {
```
Comment (Smityz) [inline src/server/pegasus_server_impl_init.cpp:57]: you can validator here is `read_amp_bytes_per_bi`t a power of 2 or equal to 0

File: src/server/pegasus_server_impl_init.cpp [hunk 4]
```diff
@@ -44,6 +44,20 @@ DSN_DEFINE_bool("pegasus.server",
...
+    return read_amp_bytes_per_bit >= 0;
```
Comment (levy5307) [inline src/server/pegasus_server_impl_init.cpp:58]: ``` return read_amp_bytes_per_bit >= 0 && read_amp_bytes_per_bit & (read_amp_bytes_per_bit - 1) == 0;```

File: src/server/pegasus_server_impl.cpp [hunk 5]
```diff
@@ -2274,6 +2288,33 @@ void pegasus_server_impl::update_replica_rocksdb_statistics()
...
+        auto write_amplification = write_amplification_iter == props.end()
+                                       ? 1
+                                       : std::stod(write_amplification_iter->second);
+        _pfc_rdb_write_amplification->set(write_amplification);
+        dinfo_replica("_pfc_rdb_write_amplification: {}", write_amplification);
+    }
+
+    // Update _pfc_rdb_read_amplification
+    if (FLAGS_read_amp_bytes_per_bit > 0) {
+        int64_t estimate_useful_bytes =
```
Comment (levy5307) [inline src/server/pegasus_server_impl.cpp:2306]: uint64_t

File: src/server/pegasus_server_impl.cpp (hunk 1 above)
Comment (Shuo-Jia) [reply src/server/pegasus_server_impl.cpp:1541]: should use

File: src/server/pegasus_server_impl.h [hunk 6]
```diff
@@ -26,6 +26,7 @@
 #include <rocksdb/options.h>
 #include <dsn/perf_counter/perf_counter_wrapper.h>
 #include <dsn/dist/replication/replication.codes.h>
+#include <dsn/utility/flags.h>
```
Comment (levy5307) [inline src/server/pegasus_server_impl.h:29]: remove this line

**Review by levy5307 — DISMISSED** [review]

File: src/server/pegasus_server_impl.cpp [hunk 7]
```diff
@@ -46,6 +46,7 @@ namespace pegasus {
 namespace server {
 
 DEFINE_TASK_CODE(LPC_PEGASUS_SERVER_DELAY, TASK_PRIORITY_COMMON, ::dsn::THREAD_POOL_DEFAULT)
+DSN_DECLARE_int32(read_amp_bytes_per_bit);
```
Comment (hycdong) [inline src/server/pegasus_server_impl.cpp:49]: I suggest this config mutable.

**Comment by Smityz** [comment]
LGTM

File: src/server/pegasus_server_impl.cpp [hunk 8]
```diff
@@ -2274,6 +2289,33 @@ void pegasus_server_impl::update_replica_rocksdb_statistics()
...
+                                       ? 1
+                                       : std::stod(write_amplification_iter->second);
+        _pfc_rdb_write_amplification->set(write_amplification);
+        dinfo_replica("_pfc_rdb_write_amplification: {}", write_amplification);
+    }
+
+    // Update _pfc_rdb_read_amplification
+    if (FLAGS_read_amp_bytes_per_bit > 0) {
+        auto estimate_useful_bytes =
+            _statistics->getTickerCount(rocksdb::READ_AMP_ESTIMATE_USEFUL_BYTES);
```
Comment (hycdong) [inline src/server/pegasus_server_impl.cpp:2308]: https://github.com/facebook/rocksdb/blob/v6.6.4/monitoring/statistics.cc
I notice that there are other statistics similar to what we like, such as `GET_HIT_L0`, `GET_HIT_L1`, `GET_HIT_L2_AND_UP`, what it the difference between them?

File: src/server/pegasus_server_impl.cpp (hunk 7 above)
Comment (Shuo-Jia) [reply src/server/pegasus_server_impl.cpp:49]: it actually pass to rocksdb, so won't be updated rocksdb though set mutable

File: src/server/pegasus_server_impl.cpp (hunk 8 above)
Comment (Shuo-Jia) [reply src/server/pegasus_server_impl.cpp:2308]: Oh, I haven't noticed, it seem hitRate under different level

Comment (hycdong) [reply src/server/pegasus_server_impl.cpp:2308]: Yes~ I think they are actually what we need, more accurate than read amplification rate

Comment (Shuo-Jia) [reply src/server/pegasus_server_impl.cpp:2308]: https://github.com/facebook/rocksdb/blob/v6.6.4/include/rocksdb/statistics.h#L101-L106 it hit count, we need calc `hitRate` via them

**Review by hycdong — APPROVED** [review]

**Review by Smityz — APPROVED** [review]

**Pull request merged**